solidity/test/isms/MultisigIsm.t.sol

```diff
@@ -5,28 +5,65 @@ import "forge-std/Test.sol";
 
 import {IMultisigIsm} from "../../contracts/interfaces/isms/IMultisigIsm.sol";
 import {TestMailbox} from "../../contracts/test/TestMailbox.sol";
-import {StaticMultisigIsmFactory} from "../../contracts/isms/multisig/StaticMultisigIsmFactory.sol";
+import {StaticMerkleRootMultisigIsmFactory} from "../../contracts/isms/multisig/StaticMultisigIsmFactory.sol";
+import {StaticMessageIdMultisigIsmFactory} from "../../contracts/isms/multisig/StaticMultisigIsmFactory.sol";
+import {MerkleRootMultisigIsmMetadata} from "../../contracts/libs/isms/MerkleRootMultisigIsmMetadata.sol";
 import {CheckpointLib} from "../../contracts/libs/CheckpointLib.sol";
+import {StaticMOfNAddressSetFactory} from "../../contracts/libs/StaticMOfNAddressSetFactory.sol";
 import {TypeCasts} from "../../contracts/libs/TypeCasts.sol";
 import {Message} from "../../contracts/libs/Message.sol";
 import {MOfNTestUtils} from "./IsmTestUtils.sol";
 
-contract MultisigIsmTest is Test {
+abstract contract AbstractMultisigIsmTest is Test {
+    using Message for bytes;
+
     uint32 constant ORIGIN = 11;
-    StaticMultisigIsmFactory factory;
+    StaticMOfNAddressSetFactory factory;
     IMultisigIsm ism;
     TestMailbox mailbox;
 
-    function setUp() public {
-        mailbox = new TestMailbox(ORIGIN);
-        factory = new StaticMultisigIsmFactory();
+    function metadataPrefix(bytes memory message)
+        internal
+        view
+        virtual
+        returns (bytes memory);
+
+    function getMetadata(
+        uint8 m,
+        uint8 n,
+        bytes32 seed,
+        bytes memory message
+    ) internal returns (bytes memory) {
+        uint32 domain = mailbox.localDomain();
+        uint256[] memory keys = addValidators(m, n, seed);
+        uint256[] memory signers = MOfNTestUtils.choose(m, keys, seed);
+        bytes32 mailboxAsBytes32 = TypeCasts.addressToBytes32(address(mailbox));
+        bytes32 checkpointRoot = mailbox.root();
+        uint32 checkpointIndex = uint32(mailbox.count() - 1);
+        bytes32 messageId = message.id();
+        bytes32 digest = CheckpointLib.digest(
+            domain,
+            mailboxAsBytes32,
+            checkpointRoot,
+            checkpointIndex,
+            messageId
+        );
+        bytes memory metadata = abi.encodePacked(
```

Comment:
nit: Consider putting mailboxAsBytes32 in the prefix, even though it would mean duplication of one line of code